### DISCUSSION TITLE
Why are dots in user part of gmail.com addresses removed?

### DISCUSSION THREAD

**Discussion by EasyNT** [rspamd/rspamd#4726] [Q&A]
The script rspamd/lualib/lua_util.lua has:

```
 172 exports.remove_email_aliases = function(email_addr)
 173   local function check_gmail_user(addr)
 174     -- Remove all points
 175     local no_dots_user = string.gsub(addr.user, '%.', '')
 176     local cap, pluses = string.match(no_dots_user, '^([^%+][^%+]*)(%+.*)$')
 177     if cap then
 178       return cap, rspamd_str_split(pluses, '+'), nil
 179     elseif no_dots_user ~= addr.user then
 180       return no_dots_user,{},nil
 181     end
 182 
 183     return nil
 184   end
```
What was the idea behind this function? It breaks PER_USER black and white lists as a sender from gmail.com with a dot in the user part never hits a redis query.

Can I simply alter this function or why is there a no_dots_user needed?

Thanks,
Juerg

**Comment by fatalbanana** [comment]
It was added with [MSBL EBL](https://msbl.org/ebl-design.html) foremost in mind.

I'd suggest to store normalized addresses in redis (without dots in case of gmail).

Supposing the configuration in question is implemented with [multimap](https://rspamd.com/doc/modules/multimap.html) the opposite approach is to use a  [selector](https://rspamd.com/doc/configuration/selectors.html) to access original from address: e.g. `from("smtp","orig")`.

**Reply by EasyNT** [reply]
Thanks so much for your answer. The addresses get stored in redis via a sieve script that pipes the sender to a bash script that then stores in redis... When you write "in case of gmail", are there any other cases with normalized addresses I should take care of and how can I find all such cases?

**Reply by fatalbanana** [reply]
Normalization performed by this function is [as described by EBL](https://msbl.org/ebl-design.html) (so: it's particular to gmail/googlemail).

**Reply by fatalbanana** [reply]
Rather, additionally: plus addresses are normalized for **all** domains.,

**Reply by EasyNT** [reply]
Ok, then I'll write a function that normalizes the addresses accordingly before stored in redis.

**Comment by vstakhov** [comment]
Please bear in mind that it is utterly stupid idea to store unnormalised addresses in any sort of list: a spammer can add any number of dots/plus signs using a single compromised account. Rspamd does the proper thing here.

**Reply by EasyNT** [reply]
Good to know, thanks!

**Reply by vstakhov** [reply]
We probably need some sort of script in rspamadm (?) to do this work to assist external tools/scripts when filling the lists.

**Reply by EasyNT** [reply]
Can I in the meantime figure out what other cases of normalisation are needed? Are there more cases than gmail?
